ui/ui-expenses/src/store/drivers/pica.js

```diff
@@ -0,0 +1,50 @@
+/* eslint-env browser */
+import pica from 'pica/dist/pica'
+
+const makeReducingCanvas = (img, coeff = 1) => {
+  const { width, height } = img
+  // if image it's too small, coeff is reduced
+  const coeffAdjusted = Math.min(Math.max(width, height), 1000) * (coeff / 1000)
+  // make canvas with reduced size
+  const canvas = window.document.createElement('canvas')
+  canvas.width = width / coeffAdjusted
+  canvas.height = height / coeffAdjusted
+  canvas.getContext('2d').drawImage(img, 0, 0, canvas.width, canvas.height)
+  return canvas
+}
+
+const reduceImg = (img, coeff) => new Promise(async (resolve, reject) => {
+  // Create an abstract resizing canvas
+  const canvas = makeReducingCanvas(img, coeff)
+  // zip image
+  const picaRunner = pica()
+  const result = await picaRunner.resize(img, canvas)
+  const blob = await picaRunner.toBlob(result, 'image/jpeg')
+  // Blob to file
+  const reader = new window.FileReader()
+  reader.readAsDataURL(blob)
+  reader.onload = readerE => resolve(readerE.target.result)
+  reader.onerror = reject
+})
+
+const reduce = store => (file, coeff) => new Promise(async (resolve) => {
+  // Create an image with input file
+  const img = window.document.createElement('img')
+  img.src = window.URL.createObjectURL(file)
+  // Once the image loads, render the image on the canvas
+  img.onload = async () => {
+    // Execute callback with the base64 URI of the image
+    const image = (await reduceImg(img, coeff)).replace(/data:.*;base64,/, '')
+    // send an event with reduced image
+    store.dispatch({ type: '@@pica/IMAGE_REDUCED', payload: image })
+    // return the reduced image
+    resolve(image)
+  }
+  img.onerror = error => store.dispatch({ type: '@@pica/IMAGE_REDUCED>ERROR', payload: error })
```

Comment:
If you don't use event to communicate with the driver (as you say in one of the comment).
Then the API should not use events, and you should reject the promise here with the given error.

(But in my mind this is better to use events...)

Reply:
no, if you have an error on this, an event it's more usable.

Reply:
When I write a driver, I'm thinking about reuse on another application.
On k-mille, event driver is not necessary, but on other application, I think it is.

So, I'm to let event AND promise resolve.

Reply:
When I write a driver, I'm thinking about reuse on another application.
 1. This not how we should code since it takes time for almost nothing the first time you write your implementation
 2. This is not the case here because this is so close to our precise need: who will need something that reduce image to 1000px max, in JPEG from a given data. You could wrote a pica driver (a real one) but it would take too much time atm and you don't use it, so what is the point ?

And, once again, I am not asking you to remove  event base driver. I am asking you to be consistent because right know this is not the case and this is "hard" to review/read so it will be "hard" to maintain **plus** we almost right event base driver when we need async (http is a good example), so why do you want to change here and let us getting lost in this code base ?

Reply:
If you prefer, rename your driver "imageResize" or something like that, because you are right this is not a pica driver, this is not something reusable.

Reply:
ok ...